src/parse/tests.rs

```diff
@@ -0,0 +1,39 @@
+use crate::ast::SrcSpan;
+use crate::parse::error::{LexicalError, LexicalErrorType, ParseError, ParseErrorType};
+
+macro_rules! assert_error {
+    ($src:expr, $error:expr $(,)?) => {
+        let result = crate::parse::parse_expression_sequence($src).expect_err("should not parse");
+        assert_eq!(($src, $error), ($src, result),);
+    };
+}
+
+#[test]
+fn bad_radix() {
+    // bad binary digit
+    assert_error!(
+        "0b012",
+        ParseError {
+            error: ParseErrorType::LexError {
+                error: LexicalError {
+                    error: LexicalErrorType::DigitOutOfRadix,
+                    location: 4,
+                }
+            },
+            location: SrcSpan { start: 4, end: 4 },
+        }
+    );
+    // bad octal digit
+    assert_error!(
+        "0o12345678",
+        ParseError {
+            error: ParseErrorType::LexError {
+                error: LexicalError {
+                    error: LexicalErrorType::DigitOutOfRadix,
+                    location: 9,
+                }
+            },
+            location: SrcSpan { start: 9, end: 9 },
+        }
+    );
+}
```

Comment:
Added a LexError on a digit out of radix, happy to remove it if we don't really need it. Without it `0b012` gets lexed as `0b01; 2`.

Reply:
I think this is a good error to raise, it'll be clearer than the alternative as to what the problem is.

Combining the errors sounds good but I'm happy either way.